core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala

```diff
@@ -145,15 +146,23 @@ object FileCommitProtocol {
       jobId: String,
       outputPath: String,
       dynamicPartitionOverwrite: Boolean = false): FileCommitProtocol = {
+
+    logDebug(s"Creating committer $className; job $jobId; output=$outputPath;" +
+      s" dynamic=$dynamicPartitionOverwrite")
     val clazz = Utils.classForName(className).asInstanceOf[Class[FileCommitProtocol]]
     // First try the constructor with arguments (jobId: String, outputPath: String,
     // dynamicPartitionOverwrite: Boolean).
     // If that doesn't exist, try the one with (jobId: string, outputPath: String).
     try {
       val ctor = clazz.getDeclaredConstructor(classOf[String], classOf[String], classOf[Boolean])
+      logDebug("Using (String, String, Boolean) constructor")
       ctor.newInstance(jobId, outputPath, dynamicPartitionOverwrite.asInstanceOf[java.lang.Boolean])
     } catch {
       case _: NoSuchMethodException =>
+        logDebug("Falling back to (String, String) constructor")
+        require(!dynamicPartitionOverwrite,
+          "Dynamic Partition Overwrite is enabled but" +
+            s" the committer ${className} does not have the appropriate constructor")
```

Comment:
BTW, why don't we warn and continue? Just wanted to make sure that we took this case into account. For example, 
wouldn't this invalidate the case below?

```scala
private class CommitProtocol(arg1: String, arg2: String)
  extends HadoopMapReduceCommitProtocol(arg1, arg2, true) {
}
```

Reply:
Problem is that the dynamic partition logic in `InsertIntoHadoopFsRelationCommand` assumes that rename() is a fast reliable operation you can do with any implementation of the FileCommitProtocol,  sets itself up for it when enabled, then instantiates the inner committer, and carries on with the dynamic partitioning, irrespective of whether or not. rename() doesn't always work like that, breaking the rest of the algorithm.

If the committer doesn't have that 3-arg constructor, you can't be confident that you can do that. To silently log and continue is to run the risk that the underlying committers commit algorithm isn't compatible with the algorithm.

A fail-fast ensures that when the outcome is going to be unknown, you aren' t left trying to work out what's happened.

Regarding your example, yes, it's in trouble. Problem is: how to differentiate that from subclasses which don't know anything at all about the new feature. you can't even look for an interface on the newly created object if the base class implements it; you are left with some dynamic probe of the instance.

Reply:
Hm .. actually we could do .. for example .. 

```scala
abstract class FileCommitProtocol {
  ...
  def dynamicPartitionOverwrite: Boolean = false
}
```

```scala
class HadoopMapReduceCommitProtocol(
    jobId: String,
    path: String,
    override val dynamicPartitionOverwrite: Boolean = false)
```

(^ it's not double checked closely, for example, if the signature is safe or not. Was just an idea)

and use `committer.dynamicPartitionOverwrite` in `InsertIntoHadoopFsRelationCommand` to respect if the commit protocol supports or not, if I understood all correctly, and then produce a warning saying dynamic partition overwrite will be ignored.

_However_, sure. I think this case is kind of a made-up case and should be a corner case I guess. I don't want to suggest an overkill (maybe) and I think we don't have to make this complicated too much for now.

I am okay as is. Just wanted to make sure that we considered and checked other possible stories.

Reply:
something like that would work, though it'd be bit more convoluted...InsertIntoFSRelation would have to check, and then handle the situation of missing support.

One thing to consider in any form is: all implementations of FileCommitProtocol should be aware of the new Dynamic Partition overwrite feature...adding a new 3-arg constructor is an implicit way of saying "I understand this". Where it's weak is there's no way for for it to say "I understand this and will handle it myself" Because essentially that's what being done in the [Netflix Partioned committer(https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/staging/PartitionedStagingCommitter.java#L142), which purges all parts for which the new job has data. With that committer, if the insert asks for the feature then the FileCommitProtocol binding to it could (somehow) turn this on and so handle everything internally. 

Like I said, a more complex model. It'd need changes a fair way through things and then the usual complexity of getting commit logic.